### PR TITLE
[FLINK-9799] Generalize and unify state meta info snapshot

### PR DESCRIPTION
**Pull request by StefanRRichter**
## What is the purpose of the change

This PR generalizes and unifies the de/serialization of state meta information in backends. We replace the snapshots and reader/writers of the individual state types with a general `StateMetaInfoSnapshot` and the corresponding `StateMetaInfoSnapshotReadersWriters`. Backwards compatibility is maintained.

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by StefanRRichter** [comment]
CC @azagrebin

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyedBackendSerializationProxy.java [hunk 1]
```diff
@@ -98,8 +103,7 @@ public int getVersion() {
 
 	@Override
 	public int[] getCompatibleVersions() {
-		// we are compatible with version 3 (Flink 1.3.x) and version 1 & 2 (Flink 1.2.x)
-		return new int[] {VERSION, 3, 2, 1};
+		return new int[]{VERSION, 4, 3, 2, 1};
```
Comment (sihuazhou) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyedBackendSerializationProxy.java]: nit: miss a ' ' between '[]' and '{'.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/metainfo/StateMetaInfoSnapshotReadersWriters.java [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+		if (readVersion == CURRENT_STATE_META_INFO_SNAPSHOT_VERSION) {
+			// latest version shortcut
+			return CurrentReaderImpl.INSTANCE;
+		}
+
+		if (readVersion > CURRENT_STATE_META_INFO_SNAPSHOT_VERSION) {
+			throw new IllegalArgumentException("Unsupported read version for state meta info: " + readVersion);
+		}
+
+		switch (stateTypeHint) {
```
Comment (azagrebin) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/metainfo/StateMetaInfoSnapshotReadersWriters.java]: One small suggestion here is to move all the legacy stuff into separate package/class and leave here only call to legacy reader factory. That would unload this class a bit and simplify cleaning of legacy stuff.

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/SerializationProxiesTest.java [hunk 3]
```diff
@@ -126,11 +131,11 @@ public void testKeyedBackendSerializationProxyRoundtripWithSerializerSerializati
 		Assert.assertTrue(serializationProxy.getKeySerializer() instanceof UnloadableDummyTypeSerializer);
 		Assert.assertEquals(keySerializer.snapshotConfiguration(), serializationProxy.getKeySerializerConfigSnapshot());
 
-		for (RegisteredKeyedBackendStateMetaInfo.Snapshot<?, ?> meta : serializationProxy.getStateMetaInfoSnapshots()) {
-			Assert.assertTrue(meta.getNamespaceSerializer() instanceof UnloadableDummyTypeSerializer);
-			Assert.assertTrue(meta.getStateSerializer() instanceof UnloadableDummyTypeSerializer);
-			Assert.assertEquals(namespaceSerializer.snapshotConfiguration(), meta.getNamespaceSerializerConfigSnapshot());
-			Assert.assertEquals(stateSerializer.snapshotConfiguration(), meta.getStateSerializerConfigSnapshot());
+		for (StateMetaInfoSnapshot meta : serializationProxy.getStateMetaInfoSnapshots()) {
```
Comment (azagrebin) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/SerializationProxiesTest.java]: Change like
`RegisteredKeyedBackendStateMetaInfo.Snapshot<?, ?> meta -> RegisteredKeyedBackendStateMetaInfo(StateMetaInfoSnapshot)`
and then using original assertions might be more complete version of test.
Also for other meta wrappers in further tests

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/metainfo/StateMetaInfoSnapshotReadersWriters.java (hunk 2 above)
Comment (StefanRRichter) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/metainfo/StateMetaInfoSnapshotReadersWriters.java]: 👍

**Comment by StefanRRichter** [comment]
@sihuazhou @azagrebin thanks guys for the fast reviews! Will address the comments and merge.

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/SerializationProxiesTest.java (hunk 3 above)
Comment (StefanRRichter) [reply flink-runtime/src/test/java/org/apache/flink/runtime/state/SerializationProxiesTest.java]: 👍 Good point

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyedBackendSerializationProxy.java (hunk 1 above)
Comment (StefanRRichter) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyedBackendSerializationProxy.java]: Both styles are ok and used in Flink, so I will stick to this.

**Pull request closed without merging**